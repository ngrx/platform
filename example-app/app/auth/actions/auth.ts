import { Action } from '@ngrx/store';
import { User, Authenticate } from '../models/user';

export enum AuthActionTypes {
  Login = '[Auth] Login',
  Logout = '[Auth] Logout',
  LoginSuccess = '[Auth] Login Success',
  LoginFailure = '[Auth] Login Failure',
  LoginRedirect = '[Auth] Login Redirect',
}

export class Login implements Action {
  readonly type = AuthActionTypes.Login;

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LoginSuccess;

  constructor(public payload: { user: User }) {}
}

export class LoginFailure implements Action {
  readonly type = AuthActionTypes.LoginFailure;

  constructor(public payload: any) {}
}

export class LoginRedirect implements Action {
  readonly type = AuthActionTypes.LoginRedirect;
}

export class Logout implements Action {
  readonly type = AuthActionTypes.Logout;
}

export type AuthActions =
  | Login
  | LoginSuccess
  | LoginFailure
  | LoginRedirect
  | Logout;
