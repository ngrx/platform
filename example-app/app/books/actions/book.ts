import { Action } from '@ngrx/store';
import { Book } from '../models/book';

export const SEARCH = '[Book] Search';
export const SEARCH_COMPLETE = '[Book] Search Complete';
export const LOAD = '[Book] Load';
export const SELECT = '[Book] Select';

/**
 * Every action is comprised of at least a type and an optional
 * payload. Expressing actions as classes enables powerful
 * type checking in reducer functions.
 *
 * See Discriminated Unions: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions
 */
export class Search implements Action {
  readonly type = SEARCH;

  constructor(public payload: string) {}
}

export class SearchComplete implements Action {
  readonly type = SEARCH_COMPLETE;

  constructor(public payload: Book[]) {}
}

export class Load implements Action {
  readonly type = LOAD;

  constructor(public payload: Book) {}
}

export class Select implements Action {
  readonly type = SELECT;

  constructor(public payload: string) {}
}

/**
 * Export a type alias of all actions in this action group
 * so that reducers can easily compose action types
 */
export type Actions = Search | SearchComplete | Load | Select;
