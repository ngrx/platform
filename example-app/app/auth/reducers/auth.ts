import * as auth from '../actions/auth';
import { User } from '../models/user';

export interface State {
  loggedIn: boolean;
  user: User | null;
}

export const initialState: State = {
  loggedIn: false,
  user: null,
};

export function reducer(state = initialState, action: auth.Actions): State {
  switch (action.type) {
    case auth.LOGIN_SUCCESS: {
      return {
        ...state,
        loggedIn: true,
        user: action.payload.user,
      };
    }

    case auth.LOGOUT: {
      return initialState;
    }

    default: {
      return state;
    }
  }
}

export const getLoggedIn = (state: State) => state.loggedIn;
export const getUser = (state: State) => state.user;
