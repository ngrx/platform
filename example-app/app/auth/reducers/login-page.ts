import * as auth from '../actions/auth';

export interface State {
  error: string | null;
  pending: boolean;
}

export const initialState: State = {
  error: null,
  pending: false,
};

export function reducer(state = initialState, action: auth.Actions): State {
  switch (action.type) {
    case auth.LOGIN: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }

    case auth.LOGIN_SUCCESS: {
      return {
        ...state,
        error: null,
        pending: false,
      };
    }

    case auth.LOGIN_FAILURE: {
      return {
        ...state,
        error: action.payload,
        pending: false,
      };
    }

    default: {
      return state;
    }
  }
}

export const getError = (state: State) => state.error;
export const getPending = (state: State) => state.pending;
