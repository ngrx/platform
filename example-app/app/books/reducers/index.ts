import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromSearch from './search';
import * as fromBooks from './books';
import * as fromCollection from './collection';
import * as fromRoot from '../../reducers';

export interface BooksState {
  search: fromSearch.State;
  books: fromBooks.State;
  collection: fromCollection.State;
}

export interface State extends fromRoot.State {
  books: BooksState;
}

export const reducers = {
  search: fromSearch.reducer,
  books: fromBooks.reducer,
  collection: fromCollection.reducer,
};

/**
 * A selector function is a map function factory. We pass it parameters and it
 * returns a function that maps from the larger state tree into a smaller
 * piece of state. This selector simply selects the `books` state.
 *
 * Selectors are used with the `select` operator.
 *
 * ```ts
 * class MyComponent {
 * 	constructor(state$: Observable<State>) {
 * 	  this.booksState$ = state$.select(getBooksState);
 * 	}
 * }
 * ```
 */

/**
 * The createFeatureSelector function selects a piece of state from the root of the state object.
 * This is used for selecting feature states that are loaded eagerly or lazily.
 */
export const getBooksState = createFeatureSelector<BooksState>('books');

/**
 * Every reducer module exports selector functions, however child reducers
 * have no knowledge of the overall state tree. To make them usable, we
 * need to make new selectors that wrap them.
 *
 * The createSelector function creates very efficient selectors that are memoized and
 * only recompute when arguments change. The created selectors can also be composed
 * together to select different pieces of state.
 */
export const getBookEntitiesState = createSelector(
  getBooksState,
  state => state.books
);

export const getSelectedBookId = createSelector(
  getBookEntitiesState,
  fromBooks.getSelectedId
);

/**
 * Adapters created with @ngrx/entity generate
 * commonly used selector functions including
 * getting all ids in the record set, a dictionary
 * of the records by id, an array of records and
 * the total number of records. This reducers boilerplate
 * in selecting records from the entity state.
 */
export const {
  selectIds: getBookIds,
  selectEntities: getBookEntities,
  selectAll: getAllBooks,
  selectTotal: getTotalBooks,
} = fromBooks.adapter.getSelectors(getBookEntitiesState);

export const getSelectedBook = createSelector(
  getBookEntities,
  getSelectedBookId,
  (entities, selectedId) => {
    return selectedId && entities[selectedId];
  }
);

/**
 * Just like with the books selectors, we also have to compose the search
 * reducer's and collection reducer's selectors.
 */
export const getSearchState = createSelector(
  getBooksState,
  (state: BooksState) => state.search
);

export const getSearchBookIds = createSelector(
  getSearchState,
  fromSearch.getIds
);
export const getSearchQuery = createSelector(
  getSearchState,
  fromSearch.getQuery
);
export const getSearchLoading = createSelector(
  getSearchState,
  fromSearch.getLoading
);
export const getSearchError = createSelector(
  getSearchState,
  fromSearch.getError
);

/**
 * Some selector functions create joins across parts of state. This selector
 * composes the search result IDs to return an array of books in the store.
 */
export const getSearchResults = createSelector(
  getBookEntities,
  getSearchBookIds,
  (books, searchIds) => {
    return searchIds.map(id => books[id]);
  }
);

export const getCollectionState = createSelector(
  getBooksState,
  (state: BooksState) => state.collection
);

export const getCollectionLoaded = createSelector(
  getCollectionState,
  fromCollection.getLoaded
);
export const getCollectionLoading = createSelector(
  getCollectionState,
  fromCollection.getLoading
);
export const getCollectionBookIds = createSelector(
  getCollectionState,
  fromCollection.getIds
);

export const getBookCollection = createSelector(
  getBookEntities,
  getCollectionBookIds,
  (entities, ids) => {
    return ids.map(id => entities[id]);
  }
);

export const isSelectedBookInCollection = createSelector(
  getCollectionBookIds,
  getSelectedBookId,
  (ids, selected) => {
    return ids.indexOf(selected) > -1;
  }
);
