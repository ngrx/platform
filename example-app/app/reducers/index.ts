import {
  ActionReducerMap,
  createSelector,
  createFeatureSelector,
  compose,
  ActionReducer,
  combineReducers,
  Action,
  ActionReducerFactory,
} from '@ngrx/store';
import * as fromRouter from '@ngrx/router-store';
import { environment } from '../../environments/environment';

/**
 * Every reducer module's default export is the reducer function itself. In
 * addition, each module should export a type or interface that describes
 * the state of the reducer plus any selector functions. The `* as`
 * notation packages up all of the exports into a single object.
 */

import * as fromLayout from '../core/reducers/layout';

/**
 * As mentioned, we treat each reducer like a table in a database. This means
 * our top level state interface is just a map of keys to inner state types.
 */
export interface State {
  layout: fromLayout.State;
}

/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const reducers: ActionReducerMap<State> = {
  layout: fromLayout.reducer,
};

/**
 * Layout Reducers
 */
export const getLayoutState = createFeatureSelector<fromLayout.State>('layout');

export const getShowSidenav = createSelector(
  getLayoutState,
  fromLayout.getShowSidenav
);
