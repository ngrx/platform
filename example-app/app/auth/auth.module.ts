import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterModule } from '@angular/router';
import { ReactiveFormsModule } from '@angular/forms';
import { StoreModule } from '@ngrx/store';
import { EffectsModule } from '@ngrx/effects';
import { LoginPageComponent } from './containers/login-page.component';
import { LoginFormComponent } from './components/login-form.component';

import { AuthService } from './services/auth.service';
import { AuthGuard } from './services/auth-guard.service';
import { AuthEffects } from './effects/auth.effects';
import { reducers } from './reducers';
import { MaterialModule } from '../material';

export const COMPONENTS = [LoginPageComponent, LoginFormComponent];

@NgModule({
  imports: [CommonModule, ReactiveFormsModule, MaterialModule],
  declarations: COMPONENTS,
  exports: COMPONENTS,
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [AuthService, AuthGuard],
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    RouterModule.forChild([{ path: 'login', component: LoginPageComponent }]),
    StoreModule.forFeature('auth', reducers),
    EffectsModule.forFeature([AuthEffects]),
  ],
})
export class RootAuthModule {}
