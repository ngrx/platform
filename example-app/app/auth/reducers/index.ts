import { createSelector, createFeatureSelector } from '@ngrx/store';
import * as fromRoot from '../../reducers';
import * as fromAuth from './auth';
import * as fromLoginPage from './login-page';

export interface AuthState {
  status: fromAuth.State;
  loginPage: fromLoginPage.State;
}

export interface State extends fromRoot.State {
  auth: AuthState;
}

export const reducers = {
  status: fromAuth.reducer,
  loginPage: fromLoginPage.reducer,
};

export const selectAuthState = createFeatureSelector<AuthState>('auth');

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state: AuthState) => state.status
);
export const getLoggedIn = createSelector(
  selectAuthStatusState,
  fromAuth.getLoggedIn
);
export const getUser = createSelector(selectAuthStatusState, fromAuth.getUser);

export const selectLoginPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginPage
);
export const getLoginPageError = createSelector(
  selectLoginPageState,
  fromLoginPage.getError
);
export const getLoginPagePending = createSelector(
  selectLoginPageState,
  fromLoginPage.getPending
);
