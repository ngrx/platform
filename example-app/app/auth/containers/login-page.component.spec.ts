import { TestBed, ComponentFixture } from '@angular/core/testing';
import { MatInputModule, MatCardModule } from '@angular/material';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { StoreModule, Store, combineReducers } from '@ngrx/store';
import { LoginPageComponent } from './login-page.component';
import { LoginFormComponent } from '../components/login-form.component';
import * as Auth from '../actions/auth';
import * as fromAuth from '../reducers';

describe('Login Page', () => {
  let fixture: ComponentFixture<LoginPageComponent>;
  let store: Store<fromAuth.State>;
  let instance: LoginPageComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        NoopAnimationsModule,
        StoreModule.forRoot({
          auth: combineReducers(fromAuth.reducers),
        }),
        MatInputModule,
        MatCardModule,
        ReactiveFormsModule,
      ],
      declarations: [LoginPageComponent, LoginFormComponent],
    });

    fixture = TestBed.createComponent(LoginPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.get(Store);

    spyOn(store, 'dispatch').and.callThrough();
  });

  /**
   * Container components are used as integration points for connecting
   * the store to presentational components and dispatching
   * actions to the store.
   *
   * Container methods that dispatch events are like a component's output observables.
   * Container properties that select state from store are like a component's input properties.
   * If pure components are functions of their inputs, containers are functions of state
   *
   * Traditionally you would query the components rendered template
   * to validate its state. Since the components are analagous to
   * pure functions, we take snapshots of these components for a given state
   * to validate the rendered output and verify the component's output
   * against changes in state.
   */
  it('should compile', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch a login event on submit', () => {
    const $event: any = {};
    const action = new Auth.Login($event);

    instance.onSubmit($event);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});
