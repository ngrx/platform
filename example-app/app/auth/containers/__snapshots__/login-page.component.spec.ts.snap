// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should compile 1`] = `
<bc-login-page
  error$={[Function Store]}
  pending$={[Function Store]}
  store={[Function Store]}
>
  
      <bc-login-form
    _nghost-c0=""
    ng-reflect-pending="false"
  >
    
      
    <mat-card
      _ngcontent-c0=""
      class="mat-card"
    >
      
        
      <mat-card-title
        _ngcontent-c0=""
        class="mat-card-title"
      >
        Login
      </mat-card-title>
      
        
      <mat-card-content
        _ngcontent-c0=""
        class="mat-card-content"
      >
        
          
        <form
          _ngcontent-c0=""
          class="ng-untouched ng-pristine ng-valid"
          ng-reflect-form="[object Object]"
          novalidate=""
        >
          
            
          <p
            _ngcontent-c0=""
          >
            
              
            <mat-input-container
              _ngcontent-c0=""
              class="mat-input-container mat-form-field ng-tns-c2-0 mat-primary mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid"
            >
              <div
                class="mat-input-wrapper mat-form-field-wrapper"
              >
                <div
                  class="mat-input-flex mat-form-field-flex"
                >
                  
                  <div
                    class="mat-input-infix mat-form-field-infix"
                  >
                    
                
                    <input
                      _ngcontent-c0=""
                      aria-invalid="false"
                      aria-required="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid"
                      formcontrolname="username"
                      id="mat-input-0"
                      matinput=""
                      ng-reflect-name="username"
                      ng-reflect-placeholder="Username"
                      ng-reflect-type="text"
                      placeholder="Username"
                      type="text"
                    />
                    
              
                    <span
                      class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                    >
                      
                      <label
                        aria-owns="mat-input-0"
                        class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c2-0 mat-empty mat-form-field-empty ng-star-inserted"
                        for="mat-input-0"
                        ng-reflect-ng-switch="false"
                      >
                        
                        
                        Username
                        
                        
                      </label>
                    </span>
                  </div>
                  
                </div>
                <div
                  class="mat-input-underline mat-form-field-underline"
                >
                  <span
                    class="mat-input-ripple mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
                  ng-reflect-ng-switch="hint"
                >
                  
                  
                  <div
                    class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c2-0 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                    style="opacity: 1;"
                  >
                    
                    <div
                      class="mat-input-hint-spacer mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-input-container>
            
            
          </p>
          
  
            
          <p
            _ngcontent-c0=""
          >
            
              
            <mat-input-container
              _ngcontent-c0=""
              class="mat-input-container mat-form-field ng-tns-c2-1 mat-primary mat-form-field-type-mat-input mat-form-field-can-float mat-form-field-hide-placeholder ng-untouched ng-pristine ng-valid"
            >
              <div
                class="mat-input-wrapper mat-form-field-wrapper"
              >
                <div
                  class="mat-input-flex mat-form-field-flex"
                >
                  
                  <div
                    class="mat-input-infix mat-form-field-infix"
                  >
                    
                
                    <input
                      _ngcontent-c0=""
                      aria-invalid="false"
                      aria-required="false"
                      class="mat-input-element mat-form-field-autofill-control ng-untouched ng-pristine ng-valid"
                      formcontrolname="password"
                      id="mat-input-1"
                      matinput=""
                      ng-reflect-name="password"
                      ng-reflect-placeholder="Password"
                      ng-reflect-type="password"
                      placeholder="Password"
                      type="password"
                    />
                    
              
                    <span
                      class="mat-form-field-label-wrapper mat-input-placeholder-wrapper mat-form-field-placeholder-wrapper"
                    >
                      
                      <label
                        aria-owns="mat-input-1"
                        class="mat-form-field-label mat-input-placeholder mat-form-field-placeholder ng-tns-c2-1 mat-empty mat-form-field-empty ng-star-inserted"
                        for="mat-input-1"
                        ng-reflect-ng-switch="false"
                      >
                        
                        
                        Password
                        
                        
                      </label>
                    </span>
                  </div>
                  
                </div>
                <div
                  class="mat-input-underline mat-form-field-underline"
                >
                  <span
                    class="mat-input-ripple mat-form-field-ripple"
                  />
                </div>
                <div
                  class="mat-input-subscript-wrapper mat-form-field-subscript-wrapper"
                  ng-reflect-ng-switch="hint"
                >
                  
                  
                  <div
                    class="mat-input-hint-wrapper mat-form-field-hint-wrapper ng-tns-c2-1 ng-trigger ng-trigger-transitionMessages ng-star-inserted"
                    style="opacity: 1;"
                  >
                    
                    <div
                      class="mat-input-hint-spacer mat-form-field-hint-spacer"
                    />
                  </div>
                </div>
              </div>
            </mat-input-container>
            
            
          </p>
          
  
            
          
          
  
            
          <p
            _ngcontent-c0=""
            class="loginButtons"
          >
            
              
            <button
              _ngcontent-c0=""
              mat-button=""
              type="submit"
            >
              Login
            </button>
            
            
          </p>
          
  
          
        </form>
        
        
      </mat-card-content>
      
      
    </mat-card>
    
    
  </bc-login-form>
    
</bc-login-page>
`;
